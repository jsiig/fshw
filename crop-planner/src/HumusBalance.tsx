import { PureComponent } from 'react'
import classNames from 'classnames'

type Props = {
  humusBalance: number | undefined,
  className: string | undefined,
}

type State = {
  previousHumusBalance: number | undefined,
  initial: boolean,
  direction: string | null,
  animationClassName: string,
  animationFinished: boolean,
}

export default class HumusBalance extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      previousHumusBalance: 0,
      initial: true,
      direction: null,
      animationClassName: '',
      animationFinished: true,
    };
  }

  static getDerivedStateFromProps(props: Props, state: State): State {
    if (
      typeof props.humusBalance !== 'undefined' &&
      typeof state.previousHumusBalance !== 'undefined' &&
      props.humusBalance !== state.previousHumusBalance
    ) {
      if (!state.initial) state.direction = props.humusBalance > state.previousHumusBalance ? 'up' : 'down';
      state.initial = false;
      state.previousHumusBalance = props.humusBalance;
    }

    return state;
  }

  componentDidUpdate = (prevProps: Props): void => {
    if (
      prevProps.humusBalance !== this.props.humusBalance &&
      this.state.direction &&
      !this.state.initial
    ) {
      this.startAnimating();
    }
  }

  startAnimating = (): void => {
    if (!this.state.animationFinished) {
      this.setState({
        animationClassName: '',
        animationFinished: true
      }, () => {
        // Ensure end of callstack execution to easily make sure
        // animation re-triggers even if already in progress with
        // the same css classes.
        // setTimeout 0 is a dirty hack but it works.
        setTimeout(() => this.setState({
          animationClassName: 'animating'
        }), 0);
      });
    } else {
      this.setState({
        animationClassName: 'animating',
      });
    }
  }

  onAnimationStart = (): void => {
    this.setState({
      animationFinished: false
    });
  }

  onAnimationEnd = () => {
    this.setState({
      animationFinished: true,
      animationClassName: ''
    });
  }

  getDirectionClassName = (): string => {
    if (!this.state.direction) return '';

    return 'humus-balance__direction--' + this.state.direction;
  }

  render = () =>
    <div
      className={
        classNames(
          this.getDirectionClassName(),
          this.state.animationClassName,
          this.props.className
        )
      }
      onAnimationStart={this.onAnimationStart}
      onAnimationEnd={this.onAnimationEnd}
    >
      {
        (this.props.humusBalance || 0).toFixed(2)
      }
    </div>
};
