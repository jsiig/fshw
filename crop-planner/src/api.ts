import {Crop, Field, HumusBalancesResponse, FieldsWithCropValues} from './types'

const SOIL_SERVICE_URL = 'http://localhost:3000'

export const fetchFields = async (): Promise<Array<Field>> =>
  await fetch(`${SOIL_SERVICE_URL}/fields`).then(response => response.json())

export const fetchCrops = async (): Promise<Array<Crop>> =>
  await fetch(`${SOIL_SERVICE_URL}/crops`).then(response => response.json())

export const calculateHumusBalance = async (fields_with_crop_values: FieldsWithCropValues): Promise<HumusBalancesResponse> =>
  await fetch(`${SOIL_SERVICE_URL}/humus_balance`, {
    headers: {
      'Content-Type': 'application/json'
    },
    method: 'POST',
    body: JSON.stringify({fields_with_crop_values})
  }).then(response => response.json())
