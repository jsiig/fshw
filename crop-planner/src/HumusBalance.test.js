import { shallow } from 'enzyme'

import HumusBalance from './HumusBalance'
import { fetchCrops, fetchFields } from './api'

describe('<HumusBalance />', () => {
  const humusBalance = 2;

  it('renders', async () => {
    expect(await shallow(<HumusBalance humusBalance={humusBalance} />)).toMatchSnapshot()
  });
});
