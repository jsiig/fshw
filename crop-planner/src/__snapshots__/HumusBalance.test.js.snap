// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HumusBalance /> renders 1`] = `
<div
  className=""
  onAnimationEnd={[Function]}
  onAnimationStart={[Function]}
>
  2.00
</div>
`;
